Delete and rename schema functions

eito-fis opened the pull request:
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #165, Fixes #167

Adds `delete_schema()` and `rename_schema()` functions to `db.schemas`.

**Technical details**
`delete_schema()` takes an optional `cascade` parameter, which is default `False`. When set, the schema is dropped with the cascade option, deleting all related objects. The `deletes_schema()` function also returns all tables related to the schema being dropped.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

kgodey commented:
Schema deletion should optionally take `CASCADE` or `RESTRICT` parameters, the user might want to delete everything the schema is associated with.

It would also be good to return a list of everything the schema references if deleting it with `RESTRICT` fails so that we can show the list in the frontend and ask the user if they want to delete all those objects by using `CASCADE`.

eito-fis commented:
@kgodey Question about:

>It would also be good to return a list of everything the schema references if deleting it with RESTRICT fails so that we can show the list in the frontend and ask the user if they want to delete all those objects by using CASCADE.

Is it sufficient to return the related tables as part of the error message in the response, or is there a better way of returning this information? Right now the function catches the `DependentObjectsStillExist` error, suppresses it and just returns the related tables, but this seems doesn't seem ideal.

kgodey commented:
@eito-fis That's fine for now, we can improve later if needed.

kgodey commented on db/schemas.py:
```diff
@@ -71,10 +75,65 @@ def get_all_schemas(engine):
+
+def rename_schema(schema, engine, rename_to):
+    """
+    This method deletes a Postgres schema.
```
This comment is inaccurate. :)

kgodey commented on db/schemas.py:
```diff
@@ -71,10 +75,65 @@ def get_all_schemas(engine):
+    This method deletes a Postgres schema.
+    """
+    if schema in get_all_schemas(engine):
+        with engine.begin() as connection:
```
Have you verified that this updates references to the schema from other schemas? (e.g. if there's a foreign key to a table in this schema from a different schema)

kgodey replied:
Please add a test to verify behavior.

kgodey reviewed (approved):
Looks good to me. Since this is entirely `db` focused, I'd like @mathemancer to look at it quickly before merge.

mathemancer commented on db/schemas.py:
```diff
@@ -71,10 +75,65 @@ def get_all_schemas(engine):
 
+def get_related_tables(engine, schema):
+    metadata = MetaData()
+    pg_tables = Table("pg_tables", metadata, autoload_with=engine)
+    sel = (
+        select(pg_tables.c.schemaname, pg_tables.c.tablename).
+        where(pg_tables.c.schemaname == schema)
+    )
+    with engine.begin() as conn:
+        results = conn.execute(sel).fetchall()
+
+    related_tables = [tables.reflect_table(table, schema, engine)
+                      for schema, table in results]
```
You can get these related tables by initializing the metadata with args:
```python
metadata = Metadata(bind=engine, schema=schema)
```
Then, use:
```python
metadata.reflect()
```
Then, you can get all the tables in that schema via
```python
metadata.tables
```

mathemancer reviewed (changes requested):
I'm not sure it's wise to drop a schema with dependent objects with a _partial_ list of objects to be deleted in the warning.  This would potentially give the user false confidence, and they might be confused (enraged) if they end up deleting something via a cascade that wasn't listed in the warning.  Currently, things which aren't listed are:
- types
- functions
- higher-order dependencies (e.g., tables that depend on tables in the schema)
- domains
- views
- constraints (likely not a problem)
- sequences
- indexes

There are likely more things I'm forgetting.

I suggest we avoid trying to list dependencies for the moment, and instead just display something like "some database objects depend on this schema.  Delete anyway?"  We could later add a separate `/api/v0/schemas/<schema_id>/dependencies/` endpoint that will list _all_ dependencies for a schema.  This will allow us to check them at will, and also make things like pagination more natural.  (Note that there could be hundreds or thousands of dependencies for a schema, and we'd want to avoid just returning them all in an error message blob).

kgodey commented:
@mathemancer That makes sense, and I take your point about a partial list of dependencies being confusing.

I'll create a separate issue for the list of dependencies and corresponding API. For now, @eito-fis, please update the error message to be generic as @mathemancer suggested.

eito-fis commented:
Updated based on feedback:
* Removed the `related_tables` return from `delete_schema`.
* Updated `delete_schema` to raise a `psycopg2.DependentObjectsStillExistError` instead of returning related tables. (Do we have a preference for custom errors, or are existing errors reasonable?)
* Added the `if_exists` option to `delete_schema`.
* Removed checking from `rename_schema` so it breaks as expected when targeting a schema that doesn't exist.

kgodey commented:
I think existing errors are best.

kgodey reviewed (approved):
Approved, I'll wait for @mathemancer to review/merge.

mathemancer reviewed (approved):
I think this looks good.  Thank you!
